### PR TITLE
Add check logic before delete subscription

### LINKED ISSUE #180: Check subscription if used by other when delete it
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by DanielXLee**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #180 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/controller/operandrequest/reconcile_operator.go [hunk 1]
```diff
@@ -161,8 +159,38 @@ func (r *ReconcileOperandRequest) updateSubscription(cr *operatorv1alpha1.Operan
 	return nil
 }
 
+func preDeleteCheck(operandName string, reconcileReq reconcile.Request, registryInstance *operatorv1alpha1.OperandRegistry) (int, int) {
+	klog.V(4).Infof("Pre-check for delete subscription: $s", operandName)
```
Comment (gyliu513) [inline pkg/controller/operandrequest/reconcile_operator.go]: kill `$s`

File: pkg/controller/operandrequest/reconcile_operator.go [hunk 2]
```diff
@@ -161,8 +159,38 @@ func (r *ReconcileOperandRequest) updateSubscription(cr *operatorv1alpha1.Operan
...
+	operatorStatus := registryInstance.Status.OperatorsStatus[operandName]
+	if operatorStatus.Phase != operatorv1alpha1.OperatorReady {
+		pos := registryInstance.GetReconcileRequest(operandName, reconcileReq)
```
Comment (gyliu513) [inline pkg/controller/operandrequest/reconcile_operator.go]: how about s/pos/rrs

rr means reconcile request

File: pkg/controller/operandrequest/reconcile_operator.go [hunk 3]
```diff
@@ -161,8 +159,38 @@ func (r *ReconcileOperandRequest) updateSubscription(cr *operatorv1alpha1.Operan
...
+		return nil
+	}
+	// If there are more than one reconcile request in the registry status, just update the
+	// operandrequest and operandregistry status, don't really delete the subscription.
+	if reconcileReqNum > 1 {
+		requestInstance.CleanMemberStatus(operandName)
+		if err := r.client.Status().Update(context.TODO(), requestInstance); err != nil {
+			return err
+		}
+		if err := r.deleteRegistryStatus(registryInstance, reconcileReq, operandName); err != nil {
```
Comment (gyliu513) [inline pkg/controller/operandrequest/reconcile_operator.go]: Why delete registry status here?

Comment (DanielXLee) [reply pkg/controller/operandrequest/reconcile_operator.go]: For example, iam operator requested by 2 `OperandRequest`, `ibm-common-services/common-service` and `ibm-common-services/test-req-2`, if we delete iam operator from request `ibm-common-services/test-req-2`, but actually, iam operator still used by request `ibm-common-services/common-service`, so we don't really delete this operator, just remove the status from registry cr.  When all the `reconcileRequests` deleted, this operator will be real delete.
Before delete iam operator from request `ibm-common-services/test-req-2`
```yaml
  status:
    operatorsStatus:
      ibm-iam-operator:
        phase: Running
        reconcileRequests:
        - name: common-service
          namespace: ibm-common-services
        - name: test-req-2
          namespace: ibm-common-services
```

After delete operator iam
```yaml
  status:
    operatorsStatus:
      ibm-iam-operator:
        phase: Running
        reconcileRequests:
        - name: common-service
          namespace: ibm-common-services
```

Comment (gyliu513) [reply pkg/controller/operandrequest/reconcile_operator.go]: Thanks @DanielXLee got it. 

Can you please put this in your code as comments?

File: pkg/controller/operandrequest/reconcile_operator.go (hunk 2 above)
Comment (DanielXLee) [reply pkg/controller/operandrequest/reconcile_operator.go]: looks good, I update the code to use the short name

File: pkg/controller/operandrequest/reconcile_operator.go (hunk 3 above)
Comment (DanielXLee) [reply pkg/controller/operandrequest/reconcile_operator.go]: ok

File: pkg/controller/operandrequest/reconcile_operator.go [hunk 4]
```diff
@@ -161,8 +159,61 @@ func (r *ReconcileOperandRequest) updateSubscription(cr *operatorv1alpha1.Operan
...
+	//       - name: request-2
+	//         namespace: common-services-ns
+	// After delete "mongodb-operator" from request "common-services-ns/request-1"
+	// status:
+	//   operatorsStatus:
+	//     mongodb-operator:
+	//       phase: Running
+	//       reconcileRequests:
+	//       - name: request-2
+	//         namespace: common-services-ns
```
Comment (gyliu513) [inline pkg/controller/operandrequest/reconcile_operator.go]: Thanks for adding the comments here, this is really helpful!

**Pull request merged**